File: backend/app/schemas/document.py
"""
Document metadata, versioning, sharing, lifecycle, bulk operations, and AI schemas (Milestone 11).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.models.document import (
    DocumentConfidentiality,
    DocumentLifecycleStatus,
    DocumentStatus,
    OcrStatus,
)
from app.models.document_share import DocumentSharePermission
from app.schemas.search import SearchResult

# =============================================================================
# Document Metadata Schemas
# =============================================================================


class DocumentCreate(BaseModel):
    """Payload for creating a document metadata record."""

    file_name: str = Field(..., min_length=1, max_length=500)
    storage_path: str = Field(..., min_length=1, max_length=2000)
    title: str | None = Field(default=None, max_length=500)
    description: str | None = None
    category: str | None = Field(default=None, max_length=100)
    document_type: str | None = Field(default=None, max_length=100)
    tags: list[str] = Field(default_factory=list)
    department_id: uuid.UUID | None = None
    confidentiality: DocumentConfidentiality = DocumentConfidentiality.INTERNAL
    retention_period_days: int | None = Field(default=None, ge=1)
    expires_at: datetime | None = None
    file_type: str | None = Field(default=None, max_length=100)
    file_size: int | None = Field(default=None, ge=0)
    checksum: str | None = Field(default=None, max_length=64)


class DocumentUpdate(BaseModel):
    """Payload for updating document metadata."""

    file_name: str | None = Field(default=None, min_length=1, max_length=500)
    title: str | None = Field(default=None, max_length=500)
    description: str | None = None
    category: str | None = Field(default=None, max_length=100)
    document_type: str | None = Field(default=None, max_length=100)
    tags: list[str] | None = None
    department_id: uuid.UUID | None = None
    confidentiality: DocumentConfidentiality | None = None
    retention_period_days: int | None = Field(default=None, ge=1)
    expires_at: datetime | None = None
    file_type: str | None = Field(default=None, max_length=100)
    checksum: str | None = Field(default=None, max_length=64)


class DocumentResponse(BaseModel):
    """Full enterprise document metadata response."""

    id: uuid.UUID
    file_name: str
    title: str | None = None
    description: str | None = None
    category: str | None = None
    document_type: str | None = None
    tags: list[str] = Field(default_factory=list)
    storage_path: str
    file_type: str | None = None
    file_size: int | None = None
    owner_id: uuid.UUID
    department_id: uuid.UUID | None = None
    confidentiality: DocumentConfidentiality = DocumentConfidentiality.INTERNAL
    lifecycle_status: DocumentLifecycleStatus = DocumentLifecycleStatus.ACTIVE
    status: DocumentStatus
    ocr_status: OcrStatus
    checksum: str | None = None
    retention_period_days: int | None = None
    activated_at: datetime | None = None
    archived_at: datetime | None = None
    expires_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    version_count: int = 1
    current_version_number: int = 1
    user_permission: str | None = None

    model_config = {"from_attributes": True}


class PaginatedDocumentResponse(BaseModel):
    """Paginated list envelope for document responses."""

    success: bool = True
    message: str = "Documents retrieved successfully."
    data: list[DocumentResponse]
    meta: dict[str, int]


# =============================================================================
# Version Schemas
# =============================================================================


class DocumentVersionResponse(BaseModel):
    """Single revision/version of a document."""

    id: uuid.UUID
    document_id: uuid.UUID
    version_number: int
    file_name: str
    storage_path: str
    file_type: str | None = None
    file_size: int | None = None
    checksum: str | None = None
    created_by: uuid.UUID
    is_current: bool
    change_summary: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentVersionListResponse(BaseModel):
    """List of versions for a document."""

    success: bool = True
    message: str = "Document versions retrieved successfully."
    data: list[DocumentVersionResponse]
    total_versions: int


# =============================================================================
# Share Schemas
# =============================================================================


class DocumentShareCreate(BaseModel):
    """Request payload to share a document with a user."""

    user_id: uuid.UUID = Field(..., description="UUID of grantee user.")
    permission: DocumentSharePermission = Field(
        DocumentSharePermission.VIEW,
        description="Permission level: view | download | edit | manage",
    )
    expires_at: datetime | None = Field(
        default=None,
        description="Optional expiration timestamp for the access grant.",
    )


class DocumentShareUpdate(BaseModel):
    """Request payload to update an existing share."""

    permission: DocumentSharePermission | None = None
    expires_at: datetime | None = None


class DocumentShareResponse(BaseModel):
    """Single document access grant response."""

    id: uuid.UUID
    document_id: uuid.UUID
    user_id: uuid.UUID
    user_email: str | None = None
    user_name: str | None = None
    granted_by: uuid.UUID
    grantor_name: str | None = None
    permission: DocumentSharePermission
    expires_at: datetime | None = None
    created_at: datetime
    revoked_at: datetime | None = None
    is_active: bool = True

    model_config = {"from_attributes": True}


# =============================================================================
# Bulk Operations Schemas
# =============================================================================


class BulkDocumentIdsRequest(BaseModel):
    """Base payload containing a list of document IDs."""

    document_ids: list[uuid.UUID] = Field(..., min_length=1)

    @field_validator("document_ids")
    @classmethod
    def validate_ids_not_empty(cls, v: list[uuid.UUID]) -> list[uuid.UUID]:
        if not v:
            raise ValueError("document_ids list cannot be empty.")
        return list(dict.fromkeys(v))  # deduplicate


class BulkTagRequest(BulkDocumentIdsRequest):
    """Payload to add tags to multiple documents."""

    tags: list[str] = Field(..., min_length=1)
    replace: bool = Field(default=False, description="Replace existing tags instead of appending.")


class BulkShareRequest(BulkDocumentIdsRequest):
    """Payload to share multiple documents with a single user."""

    user_id: uuid.UUID = Field(..., description="Grantee user.")
    permission: DocumentSharePermission = DocumentSharePermission.VIEW
    expires_at: datetime | None = None


class BulkOperationResultItem(BaseModel):
    """Itemized result for an individual document in a bulk operation."""

    document_id: uuid.UUID
    success: bool
    reason: str | None = None


class BulkOperationResponse(BaseModel):
    """Response envelope for bulk document operations."""

    success: bool = True
    message: str
    total: int
    succeeded: int
    failed: int
    results: list[BulkOperationResultItem]


# =============================================================================
# Document AI Schemas
# =============================================================================


class DocumentAISummaryResponse(BaseModel):
    """Summary generated by LLM for a specific document."""

    document_id: uuid.UUID
    document_name: str
    summary: str
    chunks_used: int
    sources: list[SearchResult] = Field(default_factory=list)


class DocumentAIChatRequest(BaseModel):
    """Request payload for conversational Q&A scoped to a single document."""

    message: str = Field(..., min_length=1)
    top_k: int = Field(default=5, ge=1, le=20)
    min_score: float | None = Field(default=None, ge=0.0, le=1.0)
    conversation_id: uuid.UUID | None = None


class DocumentAIChatResponse(BaseModel):
    """Conversational Q&A response scoped to a single document."""

    document_id: uuid.UUID
    document_name: str
    answer: str
    conversation_id: uuid.UUID
    sources: list[SearchResult] = Field(default_factory=list)


# =============================================================================
# Document Activity Schemas
# =============================================================================


class DocumentActivityItem(BaseModel):
    """Single sanitized audit event on a document timeline."""

    id: uuid.UUID
    action: str
    actor_id: uuid.UUID
    actor_name: str | None = None
    actor_email: str | None = None
    timestamp: datetime
    summary: str
    details: dict[str, Any] = Field(default_factory=dict)


class DocumentActivityResponse(BaseModel):
    """Paginated activity timeline response."""

    success: bool = True
    document_id: uuid.UUID
    data: list[DocumentActivityItem]
    total_events: int

File: frontend/types/document.ts
/**
 * Milestone 11 — Document Intelligence & Management TypeScript Types.
 */

export type DocumentStatus = "pending" | "processing" | "processed" | "failed";
export type OcrStatus = "pending" | "processing" | "completed" | "failed" | "skipped";

export type DocumentLifecycleStatus =
  | "draft"
  | "active"
  | "archived"
  | "expired"
  | "deleted";

export type DocumentConfidentiality =
  | "public"
  | "internal"
  | "confidential"
  | "restricted";

export type DocumentSharePermission =
  | "view"
  | "download"
  | "edit"
  | "manage";

export interface DocumentVersion {
  id: string;
  document_id: string;
  version_number: number;
  file_name: string;
  storage_path: string;
  file_type?: string | null;
  file_size?: number | null;
  checksum?: string | null;
  created_by: string;
  creator_name?: string | null;
  is_current: boolean;
  change_summary?: string | null;
  created_at: string;
}

export interface DocumentShare {
  id: string;
  document_id: string;
  user_id: string;
  user_email?: string | null;
  user_name?: string | null;
  granted_by: string;
  grantor_name?: string | null;
  permission: DocumentSharePermission;
  expires_at?: string | null;
  created_at: string;
  revoked_at?: string | null;
  is_active: boolean;
}

export interface EnterpriseDocument {
  id: string;
  file_name: string;
  title: string;
  description?: string | null;
  category?: string | null;
  document_type?: string | null;
  tags: string[];
  storage_path: string;
  file_type?: string | null;
  file_size?: number | null;
  owner_id: string;
  department_id?: string | null;
  confidentiality: DocumentConfidentiality;
  lifecycle_status: DocumentLifecycleStatus;
  status: DocumentStatus;
  ocr_status: OcrStatus;
  checksum?: string | null;
  retention_period_days?: number | null;
  activated_at?: string | null;
  archived_at?: string | null;
  expires_at?: string | null;
  created_at: string;
  updated_at: string;
  version_count: number;
  current_version_number: number;
  user_permission?: DocumentSharePermission | null;
}

export interface DocumentChunkSource {
  chunk_id: string;
  document_id: string;
  document_name: string;
  chunk_number: number;
  content: string;
  distance?: number;
  similarity?: number;
  file_type?: string | null;
  created_at?: string;
}

export interface DocumentAISummaryResponse {
  document_id: string;
  document_name: string;
  summary: string;
  chunks_used: number;
  sources: DocumentChunkSource[];
}

export interface DocumentAIChatRequest {
  message: string;
  conversation_id?: string;
  top_k?: number;
  min_score?: number;
}

export interface DocumentAIChatResponse {
  document_id: string;
  document_name: string;
  answer: string;
  conversation_id: string;
  sources: DocumentChunkSource[];
}

export interface DocumentActivityItem {
  id: string;
  action: string;
  actor_id?: string | null;
  actor_name: string;
  actor_email?: string | null;
  timestamp: string;
  summary: string;
  details: Record<string, unknown>;
}

export interface DocumentActivityResponse {
  document_id: string;
  data: DocumentActivityItem[];
  total_events: number;
}

export interface BulkOperationResultItem {
  document_id: string;
  success: boolean;
  reason?: string | null;
}

export interface BulkOperationResponse {
  message: string;
  total: number;
  succeeded: number;
  failed: number;
  results: BulkOperationResultItem[];
}

export interface DocumentQueryParams {
  page?: number;
  page_size?: number;
  search?: string;
  status?: DocumentStatus | "all";
  lifecycle_status?: DocumentLifecycleStatus | "all";
  department_id?: string;
  owner_id?: string;
  category?: string;
  document_type?: string;
  confidentiality?: DocumentConfidentiality | "all";
  tag?: string;
  shared_with_me?: boolean;
  date_from?: string;
  date_to?: string;
  sort?: string;
}
